umh-core/pkg/fsm/connection/reconcile.go

```diff
@@ -0,0 +1,368 @@
+//  3. Attempt the required state transition by sending the appropriate event.
+//
+// This function is intended to be called repeatedly (e.g. in a periodic control loop).
+// Over multiple calls, it converges the actual state to the desired state. Transitions
+// that fail are retried in subsequent reconcile calls after a backoff period.
+func (c *ConnectionInstance) Reconcile(ctx context.Context, snapshot fsm.SystemSnapshot, filesystemService filesystem.Service) (err error, reconciled bool) {
+	start := time.Now()
+	connectionInstanceName := c.baseFSMInstance.GetID()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentConnectionInstance, connectionInstanceName, time.Since(start))
+		if err != nil {
+			c.baseFSMInstance.GetLogger().Errorf("error reconciling connection instance %s: %v", connectionInstanceName, err)
+			c.PrintState()
+			// Add metrics for error
+			metrics.IncErrorCount(metrics.ComponentConnectionInstance, connectionInstanceName)
+		}
+	}()
+
+	// Check if context is already cancelled
+	if ctx.Err() != nil {
+		return ctx.Err(), false
+	}
+
+	// Step 1: If there's a lastError, see if we've waited enough.
+	if c.baseFSMInstance.ShouldSkipReconcileBecauseOfError(snapshot.Tick) {
+		err := c.baseFSMInstance.GetBackoffError(snapshot.Tick)
+		c.baseFSMInstance.GetLogger().Debugf("Skipping reconcile for Connection pipeline %s: %v", connectionInstanceName, err)
+
+		// if it is a permanent error, start the removal process and reset the error (so that we can reconcile towards a stopped / removed state)
+		if backoff.IsPermanentFailureError(err) {
+			// For permanent errors, we need special handling based on the instance's current state:
+			// 1. If already in a shutdown state (removed, removing, stopping, stopped), try force removal
+			// 2. If not in a shutdown state, attempt normal removal first, then force if needed
+			return c.baseFSMInstance.HandlePermanentError(
+				ctx,
+				err,
+				func() bool {
+					// Determine if we're already in a shutdown state where normal removal isn't possible
+					// and force removal is required
+					return c.IsRemoved() || c.IsRemoving() || c.IsStopping() || c.IsStopped()
+				},
+				func(ctx context.Context) error {
+					// Normal removal through state transition
+					return c.RemoveInstance(ctx, filesystemService)
+				},
+				func(ctx context.Context) error {
+					// Force removal when other approaches fail - bypasses state transitions
+					// and directly deletes files and resources
+					return c.service.ForceRemoveConnection(ctx, filesystemService, connectionInstanceName)
+				},
+			)
+		}
+		return nil, false
+	}
+
+	// Step 2: Detect external changes.
+	if err = c.reconcileExternalChanges(ctx, filesystemService, snapshot.Tick); err != nil {
+		// If the service is not running, we don't want to return an error here, because we want to continue reconciling
+		if !errors.Is(err, connectionsvc.ErrServiceNotExist) {
+			c.baseFSMInstance.SetError(err, snapshot.Tick)
```

Comment:
there is a little bit of unclarity here in the codebase whether `errors.Is(err, context.DeadlineExceeded)` is an error and should have SetError or not. maybe lets discuss it tomorrow